src/backend/commands/pg_profile.c

```diff
@@ -0,0 +1,788 @@
+	int 		password_grace_time = -1;
+	int		password_reuse_time = -1;
+	int		password_reuse_max = -1;
+	int		password_allow_hashed = -1;
+	DefElem		*dfailedloginattempts = NULL;
+	DefElem		*dpasswordlocktime = NULL;
+	DefElem		*dpasswordreusemax = NULL;
+
+	if (!superuser())
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+			 	 errmsg("permission denied to create profile, must be superuser")));
+
+	/* Extract options from the statement node tree */
+	foreach(option, stmt->options)
+	{
+		DefElem	*defel = (DefElem *) lfirst(option);
+
+		if (strcmp(defel->defname, "failed_login_attempts") == 0)
+		{
+			if (dfailedloginattempts)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("conflicting or redundant options"),
+						 parser_errposition(pstate, defel->location)));
+			dfailedloginattempts = defel;
+		}
+		else if (strcmp(defel->defname, "password_lock_time") == 0)
+		{
+			if (dpasswordlocktime)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("conflicting or redundant options"),
+						 parser_errposition(pstate, defel->location)));
+			dpasswordlocktime = defel;
+		}
+		else if (strcmp(defel->defname, "password_reuse_max") == 0)
+		{
+			if (dpasswordreusemax)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("conflicting or redundant options"),
+						 parser_errposition(pstate, defel->location)));
+			dpasswordreusemax = defel;
+		}
+	}
+
+	if (dfailedloginattempts && dfailedloginattempts->arg)
+	{
+		failed_login_attempts = intVal(dfailedloginattempts->arg);
+		if (failed_login_attempts == 0 ||
+			failed_login_attempts < PROFILE_UNLIMITED ||
+			failed_login_attempts > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid failed login attempts: %d", failed_login_attempts)));
+	}
+
+	if (dpasswordlocktime && dpasswordlocktime->arg)
+	{
+		password_lock_time = intVal(dpasswordlocktime->arg);
+		if (password_lock_time < PROFILE_UNLIMITED ||
+			password_lock_time > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid password lock time: %d", password_lock_time)));
+	}
+
+	if (dpasswordreusemax && dpasswordreusemax->arg)
+	{
+		password_reuse_max = intVal(dpasswordreusemax->arg);
+		if (password_reuse_max < PROFILE_UNLIMITED ||
+			password_reuse_max > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid password reuse max: %d", password_reuse_max)));
+	}
+
+	/*
+	 * Check the pg_profile relation to be certain the profile doesn't already
+	 * exist.
+	 */
+	pg_profile_rel = table_open(ProfileRelationId, RowExclusiveLock);
+	pg_profile_dsc = RelationGetDescr(pg_profile_rel);
+
+	if (OidIsValid(get_profile_oid(stmt->profile_name, true)))
+		ereport(ERROR,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("profile \"%s\" already exists",
+	    						stmt->profile_name)));
+
```

Comment:
Lack of test for this.

Reply:
The test has been included in profile.sql